### ISSUE TITLE
v6.3.0 connection bug

### ISSUE THREAD

**Issue by grod220** [penumbra-zone/web#963]
There have been reports of users having syncing issues on v6.3.0. From @conorsch on [discord](https://discord.com/channels/824484045370818580/1044457038886998137/1230552679646888026):

```
today i was able to replicate the cannot-sync problem on the web extension, using v6.3.0 from the publishes releases. here's a screenshot of the browser console error message, and a searching string: "Could not establish connection. Receiving end does not exist."

this error occurred on a fresh launch of Google Chrome, with a previously installed extension version. to resolve, i tried several things that did not work:

1. restarting the browser
2. force-refreshing the page

i was able to "kick it" and restore syncing functionality, by—drum roll please—:

1. opening the rpc settings
2. confirming the default grpc.testnet.penumbra.zone was the only rpc endpoint configured
3. pressing the Save button on the rpc endpoints, even though i made no changes
4. waiting for countdown to restart, then force-refreshed the page.

i am uncertain whether the force-refresh in step 4 was required or not.

```
<img width="1151" alt="Screenshot 2024-04-19 at 2 50 56 PM" src="https://github.com/penumbra-zone/web/assets/16624263/0a9906ea-2efe-4e87-864d-c994b6c7f296">

For minifront, this produces an "Penumbra extension unavailable" page.

**Comment by Valentine1898** [comment]
I can reproduce this, but the thing is, it's happening now because of a network halt 
From this we can deduce that you will get “Penumbra extension unavailable” if the services have not been started. When the network is halted this is related to #971
But, when @conorsch mentioned this bug, the network had not been halted yet, so we probably need to look for another reason why the services in extension might not be started 

The bad news is that the error log from minifront doesn't provide any useful information because the issue is in extension. 

If this bug does not occur for all users, we should make sure that other wallet extensions do not affect the extension transport (e.g. test in a new empty browser profile).

@conorsch it would be very helpful if you could share the extension logs after the testnet is deployed

**Comment by Valentine1898** [comment]
Based on the steps Connor described to restore sync I suspected this PR  
https://github.com/penumbra-zone/web/pull/868

1. Switched to commit `725be7783a55e784dd01ddcd01b40636fea1f6f4` (before #868 was merged) 
- Notice that local storage does not have a `grpcEndpoint` record  
![telegram-cloud-photo-size-2-5285050184096670252-y](https://github.com/penumbra-zone/web/assets/25391690/5bc5d21b-8e66-43d5-a6f8-7ca0902fb6c0)

2. Switched to commit  `71b71e82bb02a4aa1f8bed7ebf8a92d5757d251e` (after #868 merged) 
- Service worker does nothing because it is waiting for `grpcEndpoint` to be set in chrome local storage (look at [waitUntilGrpcEndpointExists()](https://github.com/penumbra-zone/web/blob/c73a8e2cac816b44540003eabe3a06ce876127ea/apps/extension/src/service-worker.ts#L49) )
- minifront shows “Penumbra extension unavailable”. 

I think that's the most likely cause of the bug 

We're probably better off with a release with a fix. I see two options: 
- fill grpcEndpoint value in local storage with default value
- show the user a mandatory dialog with grpc endpoint selection 

cc @grod220 @jessepinho

**Comment by turbocrime** [comment]
i have some work on this in #984 and #983 already

**Comment by grod220** [comment]
Ah, good find! This is tricky. Up to this point, we have made the default rpc selection on behalf of users and now that they need to select something---we have to migrate those old users somehow. Yes, think you are right Valentine. We can either a) somewhat forcefully migrate them via hardcoding a default or b) do a better job of signaling errors to users.

I'm somewhat leaning toward option B. Currently, we don't have a way of propagating errors to users within the extension:

<img width="396" alt="Screenshot 2024-04-26 at 10 41 16 AM" src="https://github.com/penumbra-zone/web/assets/16624263/3cb24f1b-6e58-40d7-92a5-92e3bf9e3aed">

However, for an immediate fix, it seems easiest to simply add a default. Perhaps we can do this with a code comment to remove before we go to mainnet. I'll create another issue for error messaging in the extension.

**Comment by Valentine1898** [comment]
It turned out to be more complicated than I anticipated 

If we just add `grpcEndpoint` to `localDefaults` again, it will break our onboarding behavior 
The service worker has a `waitUntilGrpcEndpointExists()` function that waits until the user selects `grpcEndpoint` and only then it initializes services. 
But if we just return the default value for `grpcEndpoint`, the service worker will start initializing before the user has selected `grpcEndpoint` (with the default value) 

It seems to me that we have to do a more tricky migration, which should affect users who have already finished onboarding but do not have grpcEndpoint set.

But, the thing is, we can't write the usual migration either 
But, the thing is, we can't write the usual migration either 
If we just bump the chrome local storage version to V3 and write something like
```typescript
export const migrations: Migration = {
  grpcEndpoint: {
    [LocalStorageVersion.V1]: old => old || DEFAULT_GRPC_URL,
    [LocalStorageVersion.V2]: old => old || DEFAULT_GRPC_URL,
  },
};
``` 
Then it won't work, because migrations only work when we've previously set the value 
https://github.com/penumbra-zone/web/blob/b8c82603063c753074d9e36f8e5f3b60bf154af5/packages/storage/src/chrome/base.ts#L36-L43

Also, ideally, if a new user interrupts onboarding during the RPC selection phase and resumes onboarding afterwards, we should not consider them a migration candidate

**Comment by Valentine1898** [comment]
> Also, ideally, if a new user interrupts onboarding during the RPC selection phase and resumes onboarding afterwards, we should not consider them a migration candidate

Ah, okay, we don't have to worry about that because if the user interrupts onboarding during the RPC selection phase he will get a broken extension without the grpcEndpoint installed anyway  🤔

**Comment by Valentine1898** [comment]
https://github.com/penumbra-zone/web/issues/993
